modules/shared/src/main/scala/org/tessellation/currency/dataApplication/DataApplicationTraverse.scala

```diff
@@ -15,69 +16,194 @@ import org.tessellation.security.hash.Hash
+                      }
+                  )
                 }
+          }
+        }
+
+        def cutoffPersistedCalculatedStates(ordinal: SnapshotOrdinal) =
+          logger.info(s"Cleaning persisted calculated states using logarithmic cutoff for ordinal=${ordinal.show}") >> {
+            val toKeep = cutoffLogic.cutoff(SnapshotOrdinal.MinValue, ordinal)
+
+            calculatedStateStorage.listStoredOrdinals.flatMap {
+              _.compile.toList
+                .map(_.diff(toKeep))
+                .flatMap(_.traverse(calculatedStateStorage.delete))
             }
           }
 
-          _ <- ordinal.map { lastOrdinal =>
-            dataApplication.setCalculatedState(lastOrdinal, state.calculated)
-          }.getOrElse(Applicative[F].unit)
+        def applyCache(
+          cache: Cache,
+          startingState: DataState.Base,
+          startingOrdinal: SnapshotOrdinal
+        ): F[(DataState.Base, SnapshotOrdinal)] = {
+          implicit val ordering = Order[SnapshotOrdinal].toOrdering
 
-        } yield (state, ordinal)
+          logger.info(s"Applying cache built during traversing, size=${cache.size.show}") >>
+            cache.toList.sortBy { case (ordinal, _) => ordinal }
+              .foldLeftM((startingState, startingOrdinal)) {
+                case ((state, _), (currentOrdinal, snapshot)) =>
+                  snapshot.dataApplication
+                    .map(_.blocks)
+                    .traverse {
+                      _.traverse { blockBytes =>
+                        dataApplication.deserializeBlock(blockBytes).flatMap(_.liftTo[F])
+                      }
+                    }
+                    .map(_.toList.flatten)
+                    .map(_.flatMap(_.updates.toList))
+                    .flatMap(dataApplication.combine(state, _))
+                    .flatTap {
+                      case DataState(_, calculatedState) =>
+                        logger.info(s"Persisting calculated state for ordinal=${currentOrdinal.show}") >>
+                          calculatedStateStorage.write(currentOrdinal, calculatedState)(dataApplication.serializeCalculatedState(_)) >>
+                          cutoffPersistedCalculatedStates(currentOrdinal)
+                    }
+                    .map((_, currentOrdinal))
+              }
+        }
+
+        discover >>= {
+          case (cache, Some((state, ordinal))) =>
+            logger.info(s"Discovered calculated state at ordinal=${ordinal.show}") >>
+              applyCache(cache, state, ordinal) >>= {
+              case (latestState, latestOrdinal) =>
+                dataApplication.setCalculatedState(latestOrdinal, latestState.calculated) >>
+                  calculatedStateStorage.deleteAbove(latestOrdinal) >>
+                  (state, ordinal).some.pure[F]
+            }
+          case (_, _) => none[(DataState.Base, SnapshotOrdinal)].pure[F]
+        }
       }
 
       private def fetchCurrencySnapshots(
```

Comment:
I saw such code already in `CurrencySnapshotProcessor` but I agree that we do not have proper abstractions in the code across L0 and L1 to share the code sometimes. Anyway, worth considering getting rid of duplicates in the future.

Reply:
Good point :+1: I don't know how to change that in that moment. We would have to create some kind of service that would serve for communication with the global layer but not as a simple DAO